Add support for StreamedResponse

supersmile2009 opened the pull request:
This PR adds support for StreamedResponse by making use of PHP's output buffering functions.

It's designed to behave as close as possible to original StreamedResponse handling when running Symfony with php-fpm. Compiler pass replaces Symfony's Symfony\Component\HttpKernel\EventListener\StreamedResponseListener with a custom listener, which triggers returning response to the client. The reason listener is used instead of simply calling the response processor from the HttpKernelRequestHandler is the same as originally in Symfony - a callback in StreamedResponse may depend on some services, which depend on Request context (current Request being present in RequestStack). Also error handling that you have in your Symfony app can kick in if the callback fails, which is not the case if we process the StreamedResponse outside of the request context.

Before I start writing tests, please let me know first if the overall design of this solution looks good to you, or you'd rather take some different approach.

k911 commented:
Thanks @supersmile2009 for PR, overall I think this is a really good feature, but I'll have to take more time to review it properly because I do not like the idea of keeping swoole request/response objects in symfony's request/response objects.

supersmile2009 commented:
@k911 what about creating SwooleContext service - a simple service to contain Swoole's Request and Response objects? We could set both request and response into it from HttpKernelRequestHandler. Then this Context could be injected into StreamedResponseListener (or any other service in future if anyone needs original Swoole's request/response for any reason).
SwooleContext could implement Symfony\Contracts\Service\ResetInterface to clean up state.
The downside of this solution obviously is SwooleContext being stateful.

To be honest I'm not a big fan of either of these solutions. But at least with the current implementation with the use of request attributes there are not stateful services, so no reliance on ResetInterface and no potential side-effects. And all interaction with the attributes is done through the SwooleRequestResponseContextManager, so attribute names are encapsulated.

k911 commented:
Anyway @supersmile2009  why not just process this response in `src/Bridge/Symfony/HttpFoundation/ResponseProcessor.php` directly in `src/Bridge/Symfony/HttpKernel/HttpKernelRequestHandler.php` and just create NoOpStreamedResponseListener or disable it in container?

k911 commented:
@supersmile2009 Please check out this approach https://github.com/supersmile2009/swoole-bundle/pull/1/files

supersmile2009 commented:
> Anyway @supersmile2009 why not just process this response in `src/Bridge/Symfony/HttpFoundation/ResponseProcessor.php` directly in `src/Bridge/Symfony/HttpKernel/HttpKernelRequestHandler.php` and just create NoOpStreamedResponseListener or disable it in container?

Same reason why Symfony has StreamedResponseListener instead of sending response in index.php like it does for other types of responses. I explained it in the PR description:
> a callback in StreamedResponse may depend on some services, which depend on Request context (current Request being present in RequestStack). Also error handling that you have in your Symfony app can kick in if the callback fails, which is not the case if we process the StreamedResponse outside of the request context.


One of the changes that I did which I didn't really like was this piece in K911\Swoole\Bridge\Symfony\HttpKernel\HttpKernelRequestHandler
```
        if (!$httpFoundationResponse instanceof StreamedResponse) {
            $this->responseProcessor->process($httpFoundationResponse, $response);
        }
```
Reasons:
- What if anyone needs to add any other type of response to the exclusion list. E. g. someone needs to write to Swoole's Response using some other listener. Overall this solution is not very extensible.
- Also it's tightly and at the same time indirectly coupled to the existence of the StreamedResponseListener.
- Differentiating between response types shouldn't be a responsibility of the request handler.

So I updated that part. Please have a look at the last 2 commits. Now request processors are a stack of decorators.
HttpKernelRequestHandler gets a stack with NoOpStreamedResponseProcessor on top, which achieves the same goal as that `if` statement.
StreamedResponseListener gets a proper StreamedResponseProcessor in the stack.

k911 commented on src/Bridge/Symfony/HttpFoundation/ResponseHeadersAndCookiesProcessor.php:
```diff
@@ -0,0 +1,48 @@
+use Swoole\Http\Response as SwooleResponse;
+use Symfony\Component\HttpFoundation\Response as HttpFoundationResponse;
+
+final class ResponseHeadersAndCookiesProcessor implements ResponseProcessorInterface
```
I'd name it like `ResponseHeadersAndStatusProcessor` since cookies are represented by headers

supersmile2009 replied:
done

k911 commented on src/Bridge/Symfony/HttpFoundation/StreamedResponseProcessor.php:
```diff
@@ -0,0 +1,35 @@
+    public function process(HttpFoundationResponse $httpFoundationResponse, SwooleResponse $swooleResponse): void
+    {
+        if (!$httpFoundationResponse instanceof StreamedResponse) {
+            // TODO: Throw exception?
```
Yeah, let's throw an exception. Actually you can just use `Assertion::isInstanceOf` like here: https://github.com/k911/swoole-bundle/blob/d1363137d7cd62764fe8107b8d7a5a0cf8c22b72/src/Bridge/Symfony/Messenger/SwooleServerTaskTransportHandler.php#L26

supersmile2009 replied:
done

k911 commented on src/Bridge/Symfony/HttpFoundation/StreamedResponseProcessor.php:
```diff
@@ -0,0 +1,35 @@
+                $swooleResponse->write($payload);
+            }
+            return '';
+        }, 8192);
```
Let's make buffer output size configurable via constructor e.g.:

```php
private $bufferOutputSize;

public function __construct(int $bufferOutputSize = 8192) {
  $this->bufferOutputSize = $bufferOutputSize;
}
```

supersmile2009 replied:
done

k911 commented:
@supersmile2009 I just wanted to kindly remind you of this PR, please tell me if you need any help or don't want to continue your work.

supersmile2009 commented:
H @k911
Thanks for the reminder. I haven't had a chance to work on it lately. But I still have intention to finish it. I'm quite busy preparing for a big release at my current job right now. But I'll get back to this PR by mid September, because we want to get rid of the patch and have it merged to upstream :)

k911 commented:
@supersmile2009 Great, thanks for an update! :)

supersmile2009 commented:
@k911 PR is updated with all the requested changes.

k911 commented:
Hi, @supersmile2009 thanks. Unfortunately CI and some tests failed. I'll take a look at this closer probably next week, however feel free to update PR, too - if you know what's going on.

Webonaute commented:
anything I can help with to make this release?

k911 commented:
Hi @supersmile2009, thanks for the changes. This weekend I'll try to add some missing tests for this & finally cut a release

supersmile2009 commented:
@k911 I've fixed the failing tests and added a test for the new response handlers.
However some of the code style pipelines will keep failing and it's not related to the changes proposed in this PR.
Only latest-code-style or lowest-code-style pipeline would work, but not both at the same time. The problem is the version of php-cs-fixer installed by composer. `@Symfony:risky` rule set was changed recently (see https://github.com/FriendsOfPHP/PHP-CS-Fixer/commit/3161ebfdb7f878c92f408cfc107652f1070d43dd#diff-f3e1fe2f863ec817efb493212b57b2a2c1ecdab8741d83ceb0519dbaa7fad7abL50)
So if you run the old version of CS fixer, you have one config applied. If you run the latest version, you have a different config applied.

To mitigate that I'd recommend moving php-cs-fixer away from composer dependencies. CI should always install explicitly when needed (i. e. in code style job) and use the latest version (e. g. download it with curl). It will not only eliminate the issue of different rule sets, but also make sure nothing slips out unchecked/unfixed due to the bugs in older php-cs-fixer versions. Also it doesn't make much sense to run the php-cs-fixer with different dependency versions. In fact it can run with no dependencies installed at all, making the pipeline even faster.

k911 commented:
@supersmile2009 Thanks for your update. As I commented before I agree that cs fixer issue is definitely out of scope for this PR. Although, I cannot agree with other points listed due to various reasons:

- cache (faster docker builds, a little but..)
- dev tools are also required to be up-to-date, because as you can see different version of tool = different results, and we want to keep consistency between several php versions, swoole versions and package versions, locking tools is the easiest way to achive that
- running code style tools on CI on different pipeline versions makes sure that future contributors can use any of the supported php version/latest/lowest supported dependencies to develop new code

IMO there weren't any benefits in not running it on multiple environments except wrong impression of having faster CI process, which then could cause issues that would have to be spotted manually. Also, if a tool produces different results across all supported versions it clearly means that we support incompatible version ranges and they have to be adjusted.

Rastusik commented:
hi @k911 and @supersmile2009 :) I just found out that this feature just broke the development workload in our company. 
To be more clear - we are integrating swoole bundle into our apps and until this feature, we were used to develop our app with swoole bundle enabled (for development purposes) but we still were able to serve our requests using php-fpm (our apps are older, so we're integrating swoole step by step and it takes some time to adapt the internals of the app). Coincindently the last app we are integrating is using StreamedResponses. With the current StreamedResponseListener, the inner Symfony http listeners are coupled to swoole, which breaks our php-fpm workloads.

What woud you say about an alternative approach to this problem, decoupled from swoole? it's quite similar to the firt approach of this PR - just with the change, that I would propose to set a callback attribute to the Symfony Request (on swoole to symfony request conversion). The callback would contain the code from the current StreamedResponseProcessor and the listener would use the callback from the request, if there was one, or it would run the code from the original symfony StreamedResponseListener. 

What would you say about this? I can implement it, but at first I'd like to know if the design is ok with you

k911 commented:
Hi @Rastusik, sure please propose PR from what I understood it's seems fine. Also, due to the fact that I treat this project more like a hobby and you're using it on production I must suggest that you should fork it internally in case you need to fix issues immediately and then you could backport features also to your codebase (and of course fixes here are most welcome) 

P.S. With swoole I spend endless amount of hours trying to understood what is happening and why it isn't working especially when it comes to coroutines and testing so I think you should reconsider using it on production, I found it working quite stable only for simple stateless (JSON) API applications (without ORM) or simple web servers. So good luck! :)

Rastusik commented:
hi @k911 ok, I'll make a PR

regarding the forks, I would do it if there would be issues happening more frequently, but they aren't, and we always were able to workaround them when there were some, so I don't see a problem there

regarding production deployment - our app based on this bundle is serving cca 30 000 concurrent users daily for a year now and it was never unstable, so I'm not really afraid. Maybe what is hard to do is writing the tests, but on the app side it works well (we're still not using coroutines, but we plan to and that will be the time, when I can look at it too and help to improve the tests)
